source/adios2/toolkit/format/dataman/DataManSerializer.cpp

```diff
@@ -1098,6 +1101,86 @@ void DataManSerializer::SetDestination(const std::string &dest)
 
 std::string DataManSerializer::GetDestination() { return m_Destination; }
 
+bool DataManSerializer::StepHasMinimumBlocks(const size_t step,
+                                             const int requireMinimumBlocks)
+{
+    std::lock_guard<std::mutex> l(m_DeserializedBlocksForStepMutex);
+    auto it = m_DeserializedBlocksForStep.find(step);
+    if (it != m_DeserializedBlocksForStep.end())
+    {
+        if (it->second >= requireMinimumBlocks)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+DmvVecPtr DataManSerializer::GetEarliestStep(int64_t &currentStep,
+                                             const int requireMinimumBlocks)
+{
+    TAU_SCOPED_TIMER_FUNC();
+
+    std::lock_guard<std::mutex> l(m_DataManVarMapMutex);
+
+    bool hasStep = false;
+    size_t earliestStep = std::numeric_limits<size_t>::max();
+
+    for (const auto &i : m_DataManVarMap)
+    {
+        if (earliestStep > i.first)
+        {
+            earliestStep = i.first;
+        }
+        hasStep = true;
+    }
+
+    if (not hasStep)
+    {
+        return nullptr;
+    }
+
+    if (StepHasMinimumBlocks(earliestStep, requireMinimumBlocks))
+    {
+        currentStep = earliestStep;
+        return m_DataManVarMap[earliestStep];
+    }
+
+    return nullptr;
+}
+DmvVecPtr DataManSerializer::GetLatestStep(int64_t &currentStep,
+                                           const int requireMinimumBlocks)
+{
+    TAU_SCOPED_TIMER_FUNC();
+
+    std::lock_guard<std::mutex> l(m_DataManVarMapMutex);
+
+    bool hasStep = false;
+    size_t latestStep = 0;
+
+    for (const auto &i : m_DataManVarMap)
+    {
+        if (latestStep < i.first)
+        {
+            latestStep = i.first;
+        }
+        hasStep = true;
+    }
+
+    if (not hasStep)
```

Comment:
I just found out that C++ supports `not` as a keyword alias to `!`! (Logically this is unimpeachable, but sounds like it might be better for detecting compiler bugs.. .)